lib/Zod.ts

```diff
@@ -1,21 +1,29 @@
-import { ZodError, ZodIssue, ZodSchema, z } from "zod"
-import { Constructor } from "./Types/HelperTypes"
+import { ZodEffects, ZodError, ZodIssue, ZodSchema, ZodType, z } from "zod"
+import { Prototype } from "./Types/HelperTypes"
 
-const optionalParse = <Input, Output extends Constructor>(
-  constructor: Output,
-  parse: (input: Input) => InstanceType<Output>
+/**
+ * optionalParse creates a Zod schema that accepts either an instance of a class
+ * or raw input that can be transformed into an instance of that class.
+ *
+ * @param clazz - The class (can have public or private constructor).
+ * @param schema - A Zod schema that validates and transforms raw input into the desired Output type.
+ * @returns A Zod schema that accepts either an instance of the class or raw input.
+ */
+const optionalParse = <Input, Output extends Prototype>(
+  clazz: Output,
+  schema: ZodEffects<ZodType<Input>, Output>
```

Comment:
```suggestion
  schema: ZodType<Input, Def, Output>
```
IIRC this is the way to make it support any zod schema instead of just a handful of schemas. You'll also have to update the signature on the z namespace.